File: web/engine/encoder.js
/**
 * Italian draughts: position -> network input, and the compact action space.
 * Port of dama/encoder.py.
 *
 * INPUT: a [7, 8, 8] tensor, flattened in the same order numpy uses
 * (plane*64 + row*8 + col), canonicalized to the side to move: for Black the
 * board is rotated 180 degrees and the colors swapped, so square s becomes
 * 63 - s and the network always sees itself moving up. Planes: 0 our men,
 * 1 our kings, 2 their men, 3 their kings, 4 dark-square mask, 5 capture
 * available, 6 the no-progress counter, normalized.
 *
 * ACTION: a move is named by its FIRST step,
 *
 *     index = dark_square(32) * 8 + direction(4) * 2 + mode(2)   -> 256
 *
 * with the directions in the canonical frame and mode in {step, jump}. Two
 * capture chains that share their first jump land on the SAME index; the
 * collision is resolved where the priors are built, by splitting the
 * probability evenly among the moves that share an index.
 */
import { colorOf, isKing, BLACK, N_SQUARES, DARK_SQUARES, DARK_INDEX } from "./board.js";
import { NO_PROGRESS_DRAW } from "./game.js";
import { hasCapture } from "./moves.js";

export const IN_PLANES = 7;
export const N_DIRS = 4;
export const N_MODES = 2;
export const SLOTS = N_DIRS * N_MODES;
export const POLICY_SIZE = DARK_SQUARES.length * SLOTS;   // 32 * 8 = 256
export const PLANE_SIZE = 64;

/**
 * Direction index in the canonical frame, in the same fixed order the Python
 * encoder uses: (-1,-1), (-1,+1), (+1,-1), (+1,+1). The order is not a detail:
 * it is baked into the trained weights, so a different one would silently map
 * every move to the wrong slot.
 */
const dirIndex = (dr, dc) => (dr < 0 ? (dc < 0 ? 0 : 1) : (dc < 0 ? 2 : 3));

const flipOf = (pos) => pos.turn === BLACK;
export const canonSq = (s, flip) => (flip ? N_SQUARES - 1 - s : s);

/** Position -> Float32Array(7*8*8), ready for the network. */
export function encode(pos) {
  const planes = new Float32Array(IN_PLANES * PLANE_SIZE);
  const flip = flipOf(pos);
  const turn = pos.turn;
  for (let s = 0; s < N_SQUARES; s++) {
    const p = pos.board[s];
    if (p === 0) continue;
    const cs = canonSq(s, flip);
    const rel = colorOf(p) * turn;               // +1 ours, -1 theirs
    const king = isKing(p);
    const plane = rel === 1 ? (king ? 1 : 0) : (king ? 3 : 2);
    planes[plane * PLANE_SIZE + cs] = 1.0;
  }
  // The dark-square mask: the parity of (row + col) survives a 180-degree
  // rotation, so the same mask holds in the canonical frame.
  for (let s = 0; s < N_SQUARES; s++) {
    if (((s >> 3) + (s & 7)) % 2 === 1) planes[4 * PLANE_SIZE + s] = 1.0;
  }
  // Captures are mandatory, so whether one exists changes what every move means:
  // the plane is broadcast over the whole board, like the counter below.
  if (hasCapture(pos.board, turn)) planes.fill(1.0, 5 * PLANE_SIZE, 6 * PLANE_SIZE);
  const np = Math.min(pos.noProgress / NO_PROGRESS_DRAW, 1.0);
  planes.fill(np, 6 * PLANE_SIZE, 7 * PLANE_SIZE);
  return planes;
}

/** Compact action index of a move, in the canonical frame. */
export function movePolicyIndex(move, flip) {
  const cf = canonSq(move.frm, flip);
  const c1 = canonSq(move.path[1], flip);
  const r0 = cf >> 3, col0 = cf & 7;
  const r1 = c1 >> 3, col1 = c1 & 7;
  const dr = Math.sign(r1 - r0);
  const dc = Math.sign(col1 - col0);
  const di = dirIndex(dr, dc);
  const mode = move.isCapture ? 1 : 0;
  return DARK_INDEX[cf] * SLOTS + di * N_MODES + mode;
}

/** The legal moves and their policy indices: this is the action masking. */
export function legalActions(pos) {
  const flip = flipOf(pos);
  const moves = pos.legalMoves();
  return { moves, indices: moves.map((m) => movePolicyIndex(m, flip)) };
}

/**
 * Raw policy logits -> probabilities over the legal moves only.
 *
 * Port of evaluators.priors_from_logits, collisions included: the softmax runs
 * over the DISTINCT indices, so a logit shared by several chains is not counted
 * twice, and the probability of an index is then split evenly among them. The
 * network cannot tell those chains apart, so an even split is the only
 * principled choice -- and the priors still sum to 1.
 */
export function priorsFromLogits(pos, logits) {
  const { moves, indices } = legalActions(pos);
  if (moves.length === 0) return [];
  const uniq = [];
  for (const ix of indices) if (!uniq.includes(ix)) uniq.push(ix);
  let max = -Infinity;
  for (const ix of uniq) if (logits[ix] > max) max = logits[ix];
  let sum = 0;
  const exp = uniq.map((ix) => { const e = Math.exp(logits[ix] - max); sum += e; return e; });
  const probOf = new Map();
  uniq.forEach((ix, j) => probOf.set(ix, exp[j] / sum));
  const counts = new Map();
  for (const ix of indices) counts.set(ix, (counts.get(ix) ?? 0) + 1);
  return moves.map((m, j) => ({
    move: m,
    prior: probOf.get(indices[j]) / counts.get(indices[j]),
  }));
}

/** The legal move for a policy index, or null. */
export function indexToMove(pos, index) {
  const { moves, indices } = legalActions(pos);
  for (let i = 0; i < moves.length; i++) if (indices[i] === index) return moves[i];
  return null;
}

File: dama/encoder.py
"""
Italian draughts: position -> tensor encoding and the compact action space.

Network input: a [IN_PLANES=7, 8, 8] tensor, canonicalized to the point of view
of the side to move (for Black the board is rotated 180 degrees and the colors
are swapped: square s becomes 63 - s). Planes: 4 for the pieces (our and their
men and kings) plus 3 engineered ones: dark-square mask, capture available
(captures are mandatory), no-progress counter.

COMPACT ACTION. Instead of from(64) x to(64) = 4096 indices, about 99% of which
can never occur in draughts, a move is described by its FIRST step:

    index = dark_square(32) * 8 + direction(4) * 2 + mode(2)            -> 256

  direction in {NW, NE, SW, SE} in the CANONICAL FRAME (relative to the side to
  move: "we" advance towards row 0), mode in {step, jump}. Each direction owns
  two consecutive slots, step first and jump second: index = 8s + 2d + m.

  For a multiple capture the index describes the FIRST jump only. Two legal
  chains that share their first jump therefore fall on the SAME index: the
  collision is handled downstream by priors_from_logits (softmax over the
  distinct indices only, then the probability is split evenly among the moves
  sharing an index).

  256 is 16x smaller than 4096: fewer parameters, faster convergence, a lighter
  policy head. Light squares have no slots at all: they are not generated and
  then masked, they simply do not exist in the action space.
"""
from __future__ import annotations
import numpy as np

from .board import color_of, is_king, BLACK, N_SQUARES, DARK_SQUARES, DARK_INDEX
from .game import Position, NO_PROGRESS_DRAW
from .moves import Move, has_capture

IN_PLANES = 7
# Constant mask of the dark (playable) squares. The parity of (r+c) does not
# change under a 180-degree rotation, so it holds in the canonical frame too.
_DARK_MASK = np.array([[1.0 if (r + c) % 2 == 1 else 0.0 for c in range(8)]
                       for r in range(8)], dtype=np.float32)
N_DIRS = 4
N_MODES = 2
SLOTS = N_DIRS * N_MODES               # 8 slots per square
POLICY_SIZE = len(DARK_SQUARES) * SLOTS  # 32 * 8 = 256 (dark squares only)

# diagonal directions in a fixed order (canonical frame: "we" move up)
DIR_LIST = ((-1, -1), (-1, 1), (1, -1), (1, 1))
DIR_TO_IDX = {d: i for i, d in enumerate(DIR_LIST)}


def _flip(pos: Position) -> bool:
    return pos.turn == BLACK


def canon_sq(s: int, flip: bool) -> int:
    return (N_SQUARES - 1 - s) if flip else s


def encode(pos: Position) -> np.ndarray:
    """Position -> [7, 8, 8] tensor, canonicalized to the point of view of the
    side to move. Planes: 0 our men, 1 our kings, 2 their men, 3 their kings,
    4 dark-square mask, 5 capture available (broadcast), 6 normalized
    no-progress counter (broadcast)."""
    planes = np.zeros((IN_PLANES, 8, 8), dtype=np.float32)
    flip = _flip(pos)
    turn = pos.turn
    for s in range(N_SQUARES):
        p = pos.board[s]
        if p == 0:
            continue
        cs = canon_sq(s, flip)
        r, c = divmod(cs, 8)
        rel = color_of(p) * turn        # +1 = ours, -1 = opponent's
        king = is_king(p)
        if rel == 1:
            planes[1 if king else 0, r, c] = 1.0
        else:
            planes[3 if king else 2, r, c] = 1.0
    # --- engineered planes ---
    planes[4] = _DARK_MASK                                    # geometry
    if has_capture(pos.board, turn):                         # tactics
        planes[5] = 1.0
    planes[6] = min(pos.no_progress / NO_PROGRESS_DRAW, 1.0)  # draw rule
    return planes


def move_policy_index(move: Move, flip: bool) -> int:
    """Compact action index (canonical frame): square*8 + direction*2 + mode,
    from the FIRST step of the move."""
    cf = canon_sq(move.frm, flip)
    c1 = canon_sq(move.path[1], flip)
    r0, col0 = divmod(cf, 8)
    r1, col1 = divmod(c1, 8)
    dr = (r1 > r0) - (r1 < r0)
    dc = (col1 > col0) - (col1 < col0)
    di = DIR_TO_IDX[(dr, dc)]
    mode = 1 if move.is_capture else 0
    return DARK_INDEX[cf] * SLOTS + di * N_MODES + mode   # compact index over 32 squares


def legal_actions(pos: Position):
    """Returns (moves, indices): the legal moves and their (compact, canonical)
    policy indices. This is the ACTION MASKING."""
    flip = _flip(pos)
    moves = pos.legal_moves()
    indices = [move_policy_index(m, flip) for m in moves]
    return moves, indices


def legal_mask(pos: Position) -> np.ndarray:
    """[POLICY_SIZE] vector with 1.0 on the legal actions, 0 elsewhere."""
    mask = np.zeros(POLICY_SIZE, dtype=np.float32)
    _, indices = legal_actions(pos)
    for i in indices:
        mask[i] = 1.0
    return mask


def index_to_move(pos: Position, index: int) -> Move | None:
    """Inverse mapping: the legal Move for a policy index (or None). If several
    chains share the same first step -- rare after the capture-priority rules --
    the first one is returned (deterministic)."""
    moves, indices = legal_actions(pos)
    for m, i in zip(moves, indices):
        if i == index:
            return m
    return None
